File: frontend/src/objects/submission/SubmissionStatus.ts
/** 提交状态枚举；表示队列、运行、完成或系统失败阶段。 */
export type SubmissionStatus = 'queued' | 'running' | 'completed' | 'failed'

const supportedSubmissionStatuses = ['queued', 'running', 'completed', 'failed'] as const satisfies readonly SubmissionStatus[]

/** 判断字符串是否为受支持提交状态。 */
export function isSubmissionStatus(value: string): value is SubmissionStatus {
  /** 注意：includes 需要把待测字符串临时断言为枚举联合类型，不会改变运行时值。 */
  return supportedSubmissionStatuses.includes(value as SubmissionStatus)
}

/** 判断提交状态是否已经不会继续推进；用于轮询停止条件。 */
export function isTerminalSubmissionStatus(status: SubmissionStatus): boolean {
  switch (status) {
    case 'completed':
    case 'failed':
      return true
    case 'queued':
    case 'running':
      return false
  }
}

File: backend/src/main/scala/domains/submission/objects/SubmissionStatus.scala
package domains.submission.objects

import io.circe.{Decoder, Encoder}


/** 提交判题生命周期状态；queued/running 为非终态，completed/failed 为终态。 */
enum SubmissionStatus:
  case Queued
  case Running
  case Completed
  case Failed

/** 提交状态的 JSON/数据库字符串编解码器。 */
object SubmissionStatus:
  given Encoder[SubmissionStatus] = Encoder.encodeString.contramap(encode)
  given Decoder[SubmissionStatus] = Decoder.decodeString.emap(parse)

  /** 将外部字符串解析为提交状态。 */
  def parse(value: String): Either[String, SubmissionStatus] =
    value.trim match
      case "queued" => Right(SubmissionStatus.Queued)
      case "running" => Right(SubmissionStatus.Running)
      case "completed" => Right(SubmissionStatus.Completed)
      case "failed" => Right(SubmissionStatus.Failed)
      case _ => Left("Submission status must be one of: queued, running, completed, failed.")

  private def encode(value: SubmissionStatus): String =
    value match
      case SubmissionStatus.Queued => "queued"
      case SubmissionStatus.Running => "running"
      case SubmissionStatus.Completed => "completed"
      case SubmissionStatus.Failed => "failed"
